### PR TITLE
refactor: optimizer shouldn't assume failed rules are internal error

### PR DESCRIPTION
**Pull request by wolffcm**
# Which issue does this PR close?

This does not close an issue but is related to https://github.com/apache/arrow-datafusion/issues/6108

# Rationale for this change

In IOx we use a logical planner rule to transform plans to support time series gap filling. Sometimes, a user has submitted invalid SQL that we cannot handle. In this case we want to generate a user-facing error message from the optimizer rule. It makes sense to use something besides `DataFusion::Internal` for this case.

Downstream IOx issue: https://github.com/influxdata/influxdb_iox/issues/7330

# What changes are included in this PR?

This just makes it so that when a rule invocation returns an error, it is not wrapped in `DataFusionError::Internal`.

# Are these changes tested?

I ran the existing suite of tests, and updated a SQL logic test. This is just a change to how errors are returned from  the logical planner when `skip_failed_rules` is `false`, which is not yet the default.

# Are there any user-facing changes?

These changes may be user facing in that they affect how errors are returned to users.

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 1]
```diff
@@ -317,11 +317,11 @@ select
 ----
 08:09:10.123456789 13:14:15.123456 13:14:15.123 13:14:15
 
-query error Cannot cast string 'not a time' to value of Time64\(Nanosecond\) type\. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
+query error Cannot cast string 'not a time' to value of Time64\(Nanosecond\) type
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:320]: this seems like a definite improvement to me 👍

**Comment by alamb** [comment]
Thank you @wolffcm

File: datafusion/optimizer/src/optimizer.rs [hunk 2]
```diff
@@ -304,13 +304,8 @@ impl Optimizer {
...
                         );
                         } else {
-                            let e = DataFusionError::Internal(format!(
-                                "Optimizer rule '{}' failed due to unexpected error: {}",
-                                rule.name(),
-                                e
-                            ));
                             return Err(DataFusionError::Context(
-                                rule.name().to_string(),
+                                format!("Optimizer rule '{}' failed", rule.name(),),
```
Comment (comphead) [inline datafusion/optimizer/src/optimizer.rs:308]: 👍

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 3]
```diff
@@ -317,11 +317,11 @@ select
...
 SELECT TIME 'not a time' as time;
 
 # invalid time
-query error Cannot cast string '24:01:02' to value of Time64\(Nanosecond\) type\. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
+query error Cannot cast string '24:01:02' to value of Time64\(Nanosecond\) type
```
Comment (comphead) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:324]: I dont see the new text in tests 🤔

Comment (wolffcm) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt:324]: The text following `query error` is a regular expression, so the test will pass without all the text. I could add the first part of the message, but it wasn't there to start with.

Comment (comphead) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt:324]: nah I was wrong, it covered, and thats why tests are failing
```
---- optimizer::tests::no_skip_failing_rule stdout ----
thread 'optimizer::tests::no_skip_failing_rule' panicked at 'assertion failed: `(left == right)`
  left: `"bad rule\ncaused by\nInternal error: Optimizer rule 'bad rule' failed due to unexpected error: Error during planning: rule failed. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker"`,
 right: `"Optimizer rule 'bad rule' failed\ncaused by\nError during planning: rule failed"`', datafusion\optimizer\src\optimizer.rs:465:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Comment (wolffcm) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt:324]: Yep, I missed a unit test in the very file I changed. It's fixed now.

**Review by comphead — APPROVED** [review]
lgtm

**Review by jackwener — APPROVED** [review]
Make sense to me, thanks @wolffcm

**Pull request merged**